hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/matcher/EventVerifier.java

```diff
@@ -0,0 +1,132 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.hawkbit.repository.test.util.TestContextProvider;
+import org.junit.Assert;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.springframework.cloud.bus.event.RemoteApplicationEvent;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.event.ApplicationEventMulticaster;
+
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.Multiset;
+import com.google.common.collect.Sets;
+import com.jayway.awaitility.Awaitility;
+import com.jayway.awaitility.core.ConditionTimeoutException;
+
+/**
+ * Test rule to setup and verify the event count for a method.
+ */
+public class EventVerifier implements TestRule {
+
+    private EventCaptor eventCaptor;
+
+    @Override
+    public Statement apply(final Statement test, final Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+
+                final Optional<Expect[]> expectedEvents = getExpectationsFrom(description);
+                expectedEvents.ifPresent(events -> beforeTest());
+                try {
+                    test.evaluate();
+                    expectedEvents.ifPresent(events -> afterTest(events));
+                } finally {
+                    expectedEvents.ifPresent(listener -> removeEventListener());
+                }
+            }
+        };
+    }
+
+    private Optional<Expect[]> getExpectationsFrom(final Description description) {
+        return ofNullable(description.getAnnotation(CountEvents.class)).map(CountEvents::events);
+    }
+
+    private void beforeTest() {
+        final ConfigurableApplicationContext context = TestContextProvider.getContext();
+        eventCaptor = new EventCaptor();
+        context.addApplicationListener(eventCaptor);
+    }
+
+    private void afterTest(final Expect[] expectedEvents) {
+        verifyRightCountOfEvents(expectedEvents);
+        verifyAllEventsCounted(expectedEvents);
+    }
+
+    private void verifyRightCountOfEvents(final Expect[] expectedEvents) {
+
+        for (final Expect expectedEvent : expectedEvents) {
+            try {
+                Awaitility.await().atMost(5, SECONDS).until(() -> eventCaptor.getCountFor(expectedEvent.type()),
+                        equalTo(expectedEvent.count()));
+
+            } catch (final ConditionTimeoutException ex) {
+                Assert.fail("Did not receive the expected amount of events form " + expectedEvent.type() + " Expected: "
+                        + expectedEvent.count() + " but was: " + eventCaptor.getCountFor(expectedEvent.type()));
+            }
+        }
+    }
+
+    private void verifyAllEventsCounted(final Expect[] expectedEvents) {
```

Comment:
The chance that asynchronous events are still processed in the eventbus are very unlikely, nevertheless it's possible that after all `expected` events are correct  `verifyRightCountOfEvents` still total events are missing. So I guess we'll need a `waiter` in the `verifyAllEventsCounted` as well.